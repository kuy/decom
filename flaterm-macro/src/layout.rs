use crate::props::Props;
use crate::tag::TagTokens;
use crate::PeekValue;
use crate::{children::LayoutChildren, node::Node};
use boolinator::Boolinator;
use proc_macro2::{Ident, Span};
use quote::{quote, ToTokens};
use syn::buffer::Cursor;
use syn::parse::{Parse, ParseStream, Parser};

pub struct Layout {
    pub name: Ident,
    pub children: LayoutChildren,
    pub props: Props,
}

impl Parse for Layout {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let open = input.parse::<LayoutOpenTag>()?;
        if open.is_self_closing() {
            return Ok(Self {
                name: open.name,
                children: LayoutChildren(vec![]),
                props: open.props,
            });
        }

        let mut children: Vec<Node> = vec![];
        loop {
            if let Some(ty) = LayoutCloseTag::peek(input.cursor()) {
                if open.name == ty {
                    break;
                }
            }

            children.push(input.parse()?);
        }

        input.parse::<LayoutCloseTag>()?;

        Ok(Self {
            name: open.name,
            children: LayoutChildren(children),
            props: open.props,
        })
    }
}

impl ToTokens for Layout {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let Self {
            name,
            children,
            props,
        } = self;
        let name_str = name.to_string();
        let layout_ident = Ident::new("__flaterm_l", Span::call_site());
        let props_token_stream = props.to_token_stream();
        let props_assignment = if props_token_stream.is_empty() {
            quote!()
        } else {
            quote!(#layout_ident.props = __flaterm_p;)
        };
        let children_token_stream = children.to_token_stream();
        let children_assignment = if children_token_stream.is_empty() {
            quote!()
        } else {
            quote!(#layout_ident.children = __flaterm_c;)
        };
        tokens.extend(quote! {
            {
                let mut #layout_ident = ::flaterm::Node::new(::std::string::String::from(#name_str));
                #props_token_stream
                #props_assignment
                #children_token_stream
                #children_assignment
                #layout_ident
            }
        });
    }
}

impl Layout {
    fn peek_name(cursor: Cursor) -> (Ident, Cursor) {
        cursor.ident().unwrap()
    }
}

pub struct LayoutOpenTag {
    pub tag: TagTokens,
    pub name: Ident,
    pub props: Props,
}

impl LayoutOpenTag {
    fn is_self_closing(&self) -> bool {
        self.tag.div.is_some()
    }
}

impl Parse for LayoutOpenTag {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let (tag, inner_brackets) = TagTokens::parse_start_tag(input)?;
        let inner_parser = |input: ParseStream| {
            let name = input.parse()?;
            let props = input.parse()?;
            Ok((name, props))
        };
        let (name, props) = inner_parser.parse2(inner_brackets)?;
        Ok(Self { tag, name, props })
    }
}

pub struct LayoutCloseTag {
    _tag: TagTokens,
    _name: Ident,
}

impl Parse for LayoutCloseTag {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let (tag, content) = TagTokens::parse_end_tag(input)?;
        let content_parser = |input: ParseStream| {
            let name = input.parse()?;
            Ok((name,))
        };
        let (name,) = content_parser.parse2(content)?;
        Ok(Self {
            _tag: tag,
            _name: name,
        })
    }
}

impl PeekValue<Ident> for LayoutCloseTag {
    fn peek(cursor: Cursor) -> Option<Ident> {
        let (punct, cursor) = cursor.punct()?;
        (punct.as_char() == '<').as_option()?;

        let (punct, cursor) = cursor.punct()?;
        (punct.as_char() == '/').as_option()?;

        let (ty, cursor) = Layout::peek_name(cursor);

        let (punct, _) = cursor.punct()?;
        (punct.as_char() == '>').as_option()?;

        Some(ty)
    }
}
