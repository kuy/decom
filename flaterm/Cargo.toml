[package]
name = "flaterm"
version = "0.0.0"
authors = ["Yuki Kodama<endflow.net@gmail.com>"]
repository = "https://github.com/kuy/decom"
homepage = "https://github.com/kuy/decom"
edition = "2018"

[dependencies]
crossterm = { version = "0.23.1", optional = true }
tui = { version = "0.16.0", default-features = false, features = ["crossterm"], optional = true }

[features]
default = ["with-flat"]
with-tui = ["tui", "crossterm"]
with-flat = ["crossterm"]
