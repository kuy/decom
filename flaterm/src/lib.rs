#[derive(Debug)]
pub struct Node {
    pub name: String,
    pub children: Vec<Node>,
}

impl Node {
    pub fn new(ty: String) -> Self {
        Self {
            name: ty,
            children: vec![],
        }
    }
}
