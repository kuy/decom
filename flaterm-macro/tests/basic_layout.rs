use flaterm::Block;
use flaterm_macro::layout;

pub fn test_basic_layout() {
    layout! {
        <Block></Block>
    };

    layout! {
        <Block />
    };
}
