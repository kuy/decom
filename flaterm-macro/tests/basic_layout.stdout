#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2018::*;
#[macro_use]
extern crate std;
use flaterm::Block;
use flaterm_macro::layout;
pub fn test_basic_layout() {
    {
        let __flaterm_item = Block::new();
        __flaterm_item
    };
    {
        let __flaterm_item = Block::new();
        __flaterm_item
    };
}
